src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,243 @@
+
+            // TO-DO: We should implement support for streaming once this LSP bug is fixed:
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132601
+            return ComputeTokens(text.Lines, classifiedSpans.ToArray(), tokenTypesToIndex);
+        }
+
+        private static int[] ComputeTokens(
+            TextLineCollection lines,
+            ClassifiedSpan[] classifiedSpans,
+            Dictionary<string, int> tokenTypesToIndex)
+        {
+            using var _ = ArrayBuilder<int>.GetInstance(out var data);
+
+            // We keep track of the last line number and last start character since tokens are
+            // reported relative to each other.
+            var lastLineNumber = 0;
+            var lastStartCharacter = 0;
+
+            for (var currentClassifiedSpanIndex = 0; currentClassifiedSpanIndex < classifiedSpans.Length; currentClassifiedSpanIndex++)
+            {
+                currentClassifiedSpanIndex = ComputeNextToken(
+                    lines, ref lastLineNumber, ref lastStartCharacter, classifiedSpans,
+                    currentClassifiedSpanIndex, tokenTypesToIndex,
+                    out var deltaLine, out var startCharacterDelta, out var tokenLength,
+                    out var tokenType, out var tokenModifiers);
+
+                data.AddRange(deltaLine, startCharacterDelta, tokenLength, tokenType, tokenModifiers);
+            }
+
+            return data.ToArray();
+        }
+
+        private static int ComputeNextToken(
+            TextLineCollection lines,
+            ref int lastLineNumber,
+            ref int lastStartCharacter,
+            ClassifiedSpan[] classifiedSpans,
+            int currentClassifiedSpanIndex,
+            Dictionary<string, int> tokenTypesToIndex,
+            // Out params
+            out int deltaLineOut,
+            out int startCharacterDeltaOut,
+            out int tokenLengthOut,
+            out int tokenTypeOut,
+            out int tokenModifiersOut)
+        {
+            // Each semantic token is represented in LSP by five numbers:
+            //     1. Token line number delta, relative to the previous token
+            //     2. Token start character delta, relative to the previous token
+            //     3. Token length
+            //     4. Token type (index) - looked up in SemanticTokensLegend.tokenTypes
+            //     5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+
+            var classifiedSpan = classifiedSpans[currentClassifiedSpanIndex];
+            var linePosition = lines.GetLinePositionSpan(classifiedSpan.TextSpan).Start;
+            var lineNumber = linePosition.Line;
+            var startCharacter = linePosition.Character;
+
+            // 1. Token line number delta, relative to the previous token
+            var deltaLine = lineNumber - lastLineNumber;
+            Contract.ThrowIfTrue(deltaLine < 0);
```

Comment:
i reommend including messges.  that way if there are multiple contract calls we know which one failed.  otherwise we just get "Unexpected false in ComputeNextToken" and we don't know which was at fault (since there aren't line numbers in release).